File: mbw/src/main/java/com/mycelium/wallet/activity/MessageSigningActivity.java
/*
 * Copyright 2013, 2014 Megion Research and Development GmbH
 *
 * Licensed under the Microsoft Reference Source License (MS-RSL)
 *
 * This license governs use of the accompanying software. If you use the software, you accept this license.
 * If you do not accept the license, do not use the software.
 *
 * 1. Definitions
 * The terms "reproduce," "reproduction," and "distribution" have the same meaning here as under U.S. copyright law.
 * "You" means the licensee of the software.
 * "Your company" means the company you worked for when you downloaded the software.
 * "Reference use" means use of the software within your company as a reference, in read only form, for the sole purposes
 * of debugging your products, maintaining your products, or enhancing the interoperability of your products with the
 * software, and specifically excludes the right to distribute the software outside of your company.
 * "Licensed patents" means any Licensor patent claims which read directly on the software as distributed by the Licensor
 * under this license.
 *
 * 2. Grant of Rights
 * (A) Copyright Grant- Subject to the terms of this license, the Licensor grants you a non-transferable, non-exclusive,
 * worldwide, royalty-free copyright license to reproduce the software for reference use.
 * (B) Patent Grant- Subject to the terms of this license, the Licensor grants you a non-transferable, non-exclusive,
 * worldwide, royalty-free patent license under licensed patents for reference use.
 *
 * 3. Limitations
 * (A) No Trademark License- This license does not grant you any rights to use the Licensor’s name, logo, or trademarks.
 * (B) If you begin patent litigation against the Licensor over patents that you think may apply to the software
 * (including a cross-claim or counterclaim in a lawsuit), your license to the software ends automatically.
 * (C) The software is licensed "as-is." You bear the risk of using it. The Licensor gives no express warranties,
 * guarantees or conditions. You may have additional consumer rights under your local laws which this license cannot
 * change. To the extent permitted under your local laws, the Licensor excludes the implied warranties of merchantability,
 * fitness for a particular purpose and non-infringement.
 */

package com.mycelium.wallet.activity;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import com.mrd.bitlib.crypto.InMemoryPrivateKey;
import com.mrd.bitlib.crypto.SignedMessage;
import com.mrd.bitlib.model.Address;
import com.mrd.bitlib.model.AddressType;
import com.mrd.bitlib.model.NetworkParameters;
import com.mrd.bitlib.util.HashUtils;
import com.mrd.bitlib.util.Sha256Hash;
import com.mycelium.wallet.MbwManager;
import com.mycelium.wallet.R;
import com.mycelium.wallet.Utils;
import com.mycelium.wallet.activity.modern.Toaster;
import com.mycelium.wapi.wallet.AesKeyCipher;
import com.mycelium.wapi.wallet.KeyCipher;
import com.mycelium.wapi.wallet.single.SingleAddressAccount;

/*
todo HD: root seeds will for now not support signing directly. only support single addresses.

todo HD: instead, there can be two possibilities. select address for signing from transaction, or selecting an address from a huge list. this sucks, so lets discuss it further.

*/
public class MessageSigningActivity extends Activity {


    public static final String PRIVATE_KEY = "privateKey";
    public static final String ADDRESS_TYPE = "addressType";
    private String base64Signature;
    private String messageText;
    private NetworkParameters network;
    public static final String TEMPLATE =
           /**/"-----BEGIN BITCOIN SIGNED MESSAGE-----\n" +
           /**/"%s\n" +
           /**/"-----BEGIN BITCOIN SIGNATURE-----\n" +
           /**/"Version: Bitcoin-qt (1.0)\n" +
           /**/"Address: %s\n" +
           /**/"\n" +
           /**/"%s\n" +
           /**/"-----END BITCOIN SIGNATURE-----";

    public static void callMe(Context currentActivity, SingleAddressAccount account, AddressType addressType) {
       InMemoryPrivateKey privateKey;
       try {
          privateKey = account.getPrivateKey(AesKeyCipher.defaultKeyCipher());
       } catch (KeyCipher.InvalidKeyCipher e) {
          throw new RuntimeException(e);
       }
       callMe(currentActivity, privateKey, addressType);
    }

   public static void callMe(Context currentActivity, InMemoryPrivateKey key, AddressType addressType) {
      Intent intent = new Intent(currentActivity, MessageSigningActivity.class);
      String privKey = key.getBase58EncodedPrivateKey(MbwManager.getInstance(currentActivity).getNetwork());
      intent.putExtra(PRIVATE_KEY, privKey);
      intent.putExtra(ADDRESS_TYPE, addressType);
      currentActivity.startActivity(intent);
   }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setTitle(R.string.sign_message);
        String encoded = getIntent().getStringExtra(PRIVATE_KEY);
        final AddressType addressType = (AddressType) getIntent().getSerializableExtra(ADDRESS_TYPE);
        network = MbwManager.getInstance(this).getNetwork();
        final InMemoryPrivateKey privateKey = new InMemoryPrivateKey(encoded, network);

        setContentView(R.layout.message_signing);
        final View signButton = findViewById(R.id.btSign);
        final View copyButton = findViewById(R.id.btCopyToClipboard);
        final View shareButton = findViewById(R.id.btShare);
        final TextView signature = findViewById(R.id.signature);

        final EditText messageToSign = findViewById(R.id.etMessageToSign);
        copyButton.setVisibility(View.GONE);
        shareButton.setVisibility(View.GONE);

        signButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                signButton.setEnabled(false);
                messageToSign.setEnabled(false);
                messageToSign.setHint("");
                final ProgressDialog pd = new ProgressDialog(MessageSigningActivity.this);
                pd.setTitle(getString(R.string.signing_inprogress));
                pd.setCancelable(false);
                pd.setIndeterminate(true);
                pd.show();

                new Thread() {
                    @Override
                    public void run() {
                        messageText = messageToSign.getText().toString();
                        SignedMessage signedMessage = privateKey.signMessage(messageText);
                        base64Signature = signedMessage.getBase64Signature();
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                pd.dismiss();
                                signature.setText(base64Signature);
                                signButton.setVisibility(View.GONE);
                                copyButton.setVisibility(View.VISIBLE);
                                shareButton.setVisibility(View.VISIBLE);
                            }
                        });
                    }
                }.start();
            }
        });

        copyButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Utils.setClipboardString(base64Signature, MessageSigningActivity.this);
                new Toaster(MessageSigningActivity.this).toast(R.string.sig_copied, false);
            }
        });
        shareButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent sharingIntent = new Intent(Intent.ACTION_SEND);
                sharingIntent.setType("text/plain");
                Address address = privateKey.getPublicKey().toAddress(network, addressType);
                String body = String.format(TEMPLATE, messageText, address, base64Signature);

                sharingIntent.putExtra(android.content.Intent.EXTRA_TEXT, body);
                sharingIntent.putExtra(android.content.Intent.EXTRA_SUBJECT, getString(R.string.signed_message_subject));
                startActivity(Intent.createChooser(sharingIntent, getString(R.string.signed_message_share)));
            }
        });
    }
}

File: mbw/src/main/java/com/mycelium/wallet/activity/MessageSigningActivity.kt
package com.mycelium.wallet.activity

import android.app.Activity
import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.mycelium.wallet.MbwManager
import com.mycelium.wallet.R
import com.mycelium.wallet.Utils
import com.mycelium.wallet.activity.modern.HDSigningActivity
import com.mycelium.wallet.activity.modern.Toaster
import com.mycelium.wapi.wallet.Address
import com.mycelium.wapi.wallet.KeyCipher.InvalidKeyCipher
import com.mycelium.wapi.wallet.WalletAccount
import com.mycelium.wapi.wallet.btc.AbstractBtcAccount
import com.mycelium.wapi.wallet.btc.bip44.AddressesListProvider
import com.mycelium.wapi.wallet.btcvault.AbstractBtcvAccount
import com.mycelium.wapi.wallet.colu.ColuAccount
import com.mycelium.wapi.wallet.eth.EthAccount
import kotlinx.android.synthetic.main.message_signing.*
import java.util.*
import kotlin.concurrent.thread

class MessageSigningActivity : AppCompatActivity() {
    private var signature: String? = null
    private var messageText: String? = null
    private var address: Address? = null
    private var account: WalletAccount<*>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setTitle(R.string.sign_message)
        account = MbwManager.getInstance(this).getWalletManager(false)
            .getAccount(intent.getSerializableExtra(ACCOUNT) as UUID)!!
        address = intent.getSerializableExtra(ADDRESS) as? Address ?: account!!.receiveAddress

        setContentView(R.layout.message_signing)
        btCopyToClipboard.visibility = View.GONE
        btShare.visibility = View.GONE
        btSign.setOnClickListener {
            btSign.isEnabled = false
            etMessageToSign.isEnabled = false
            etMessageToSign.hint = ""
            val pd = ProgressDialog(this)
            pd.setTitle(getString(R.string.signing_inprogress))
            pd.setCancelable(false)
            pd.show()
            thread {
                messageText = etMessageToSign.text.toString()
                signature = account!!.signMessage(messageText!!, address)
                runOnUiThread {
                    pd.dismiss()
                    tvSignature.text = signature
                    btSign.visibility = View.GONE
                    btCopyToClipboard.visibility = View.VISIBLE
                    btShare.visibility = View.VISIBLE
                }
            }
        }
        btCopyToClipboard.setOnClickListener {
            Utils.setClipboardString(signature, this@MessageSigningActivity)
            Toaster(this@MessageSigningActivity).toast(R.string.sig_copied, false)
        }
        btShare.setOnClickListener {
            val sharingIntent = Intent(Intent.ACTION_SEND)
            sharingIntent.type = "text/plain"
            sharingIntent.putExtra(Intent.EXTRA_TEXT, getBody())
            sharingIntent.putExtra(Intent.EXTRA_SUBJECT, getSubject())
            startActivity(Intent.createChooser(sharingIntent, getChooserTitle()))
        }
    }

    companion object {
        const val PRIVATE_KEY = "privateKey"
        const val ADDRESS = "address"
        const val ACCOUNT = "account"
        const val BTC_TEMPLATE = """-----BEGIN BITCOIN SIGNED MESSAGE-----
%s
-----BEGIN BITCOIN SIGNATURE-----
Version: Bitcoin-qt (1.0)
Address: %s

%s
-----END BITCOIN SIGNATURE-----"""

        // Template comes from MyCrypto so it could be copied and easily verified there
        const val ETH_TEMPLATE = """{
  "address": "%s",
  "msg": "%s",
  "sig": "%s",
  "version": "2"
}"""

        @JvmStatic
        fun callMe(currentActivity: Context, focusedAccount: WalletAccount<*>) {
            try {
                when (focusedAccount) {
                    is AddressesListProvider<*> -> {
                        val intent = Intent(currentActivity, HDSigningActivity::class.java)
                            .putExtra("account", focusedAccount.id)
                        currentActivity.startActivity(intent)
                    }
                    else -> {
                        callMe(currentActivity, focusedAccount.id)
                    }
                }
            } catch (invalidKeyCipher: InvalidKeyCipher) {
                invalidKeyCipher.printStackTrace()
            }
        }

        @JvmStatic
        fun callMe(currentActivity: Context, address: Address?, accountId: UUID) {
            val intent = Intent(currentActivity, MessageSigningActivity::class.java)
                .putExtra(ADDRESS, address)
                .putExtra(ACCOUNT, accountId)
            currentActivity.startActivity(intent)
        }

        @JvmStatic
        fun callMe(currentActivity: Context, accountId: UUID) {
            val intent = Intent(currentActivity, MessageSigningActivity::class.java)
                .putExtra(ACCOUNT, accountId)
            currentActivity.startActivity(intent)
        }
    }

    private fun getBody() =
        when (account) {
            is EthAccount -> String.format(
                ETH_TEMPLATE,
                address,
                messageText,
                signature
            )
            else -> String.format(
                BTC_TEMPLATE,
                messageText,
                address,
                signature
            )
        }

    private fun getChooserTitle() =
        when (account) {
            is AbstractBtcAccount, is AbstractBtcvAccount, is ColuAccount -> getString(R.string.signed_message_share_btc)
            else -> getString(R.string.signed_message_share, account!!.coinType.symbol)
        }

    private fun getSubject() =
        when (account) {
            is AbstractBtcAccount, is AbstractBtcvAccount, is ColuAccount -> getString(R.string.signed_message_subject_btc)
            else -> getString(R.string.signed_message_subject, account!!.coinType.symbol)
        }
}
